execute/report/report.go

```diff
@@ -349,11 +349,68 @@ func (b *execReportBuilder) checkMessage(
 	return result, ReadyToExecute, nil
 }
 
+// we cannot have "skipped" nonces since these messages will be skipped onchain.
+// they will also end up in the inflight cache and "snoozed" until they
+// expire from the inflight cache TTL.
+func checkSkippedNonces(
+	addressCodec ccipocr3.AddressCodec,
+	execReport ccipocr3.ExecutePluginReportSingleChain,
+) error {
+	// note that the messages in the report _must_ be ordered correctly in nonce order,
+	// otherwise the report is invalid. this is because we loop through the messages in order
+	// onchain and try to execute each in turn.
+	// for that reason, we can use this map approach by checking the latest nonce for each sender,
+	// instead of accumulating the nonces in an array and searching for gaps.
+	// so this check catches incorrectly ordered messages as well as skipped nonces.
+	latestNoncePerSender := make(map[string]uint64)
+	for _, msg := range execReport.Messages {
+		if msg.Header.Nonce == 0 {
+			// nonce is zero means an out of order execution message, it doesn't affect nonce ordering.
+			// we can safely skip this check for this message.
+			continue
+		}
+
+		sender, err := addressCodec.AddressBytesToString(msg.Sender, msg.Header.SourceChainSelector)
+		if err != nil {
+			return fmt.Errorf("unable to convert sender address to string: %w, sender address: %v", err, msg.Sender)
+		}
+
+		latestNonce, ok := latestNoncePerSender[sender]
+		if !ok {
+			// first time we ever see this sender, populate the nonce.
+			latestNoncePerSender[sender] = msg.Header.Nonce
+			continue
+		} else if latestNonce == 0 {
+			// this is a bug in this code, should never happen since we check if msg.Header.Nonce
+			// is zero before this check.
+			return fmt.Errorf("assumption violation: bug in the code, latestNonce is zero when it should never be")
+		} else if msg.Header.Nonce != latestNonce+1 {
+			// we've seen this sender before and the msg.Header.Nonce is not the expected value,
+			// it should be latestNonce + 1.
+			return fmt.Errorf(
+				"skipped nonce detected for sender %s: nonce in report %d != last nonce seen %d",
+				sender, msg.Header.Nonce, latestNonce)
```

Comment:
I guess this can technically also be incorrectly ordered nonce if the `Messages` array is incorrectly built, I can do an additional check and tweak this error msg.